mv: throttle view update generation for large queries

cvybhu opened the pull request:
This is another attempt at a fix for problems with materialized views concurrency #12379.

See #12379 and #13583 for a detailed description of the problem and discussions.

This PR aims to extend the existing throttling mechanism to work with requests that internally generate a large amount of view updates, as suggested by @nyh.

The existing mechanism works in the following way:
* Client sends a request, we generate the view updates corresponding to the request and spawn background tasks which will send these updates to remote nodes
* Each background task consumes some units from the `view_update_concurrency_semaphore`, but doesn't wait for these units, it's just for tracking
* We keep track of the percent of consumed units on each node, this is called `view update backlog`.
* Before sending a response to the client we sleep for a short amount of time. The amount of time to sleep for is based on the fullness of this `view update backlog`. For a well behaved client with limited concurrency this will limit the amount of incoming requests to a manageable level.

This mechanism doesn't handle large DELETE queries. Deleting a partition is fast for the base table, but it requires us to generate a view update for every single deleted row. The number of deleted rows per single client request can be in the millions. Delaying response to the request doesn't help when a single request can generate millions of updates.

To deal with this we could treat the view update generator just like any other client and force it to wait a bit of time before sending the next batch of updates. The amount of time to wait for is calculated just like in the existing throttling code, it's based on the fullness of `view update backlogs`.

The new algorithm of view update generation looks something like this:
```c++
for(;;) {
    auto updates = generate_updates_batch_with_max_100_rows();
    co_await seastar::sleep(calculate_sleep_time_from_backlogs());
    spawn_background_tasks_for_updates(updates);
}
```

Fixes: https://github.com/scylladb/scylladb/issues/12379

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1074/

cvybhu commented:
Got a weird error in the CI:
```
22:17:47  [2941/3007] scripts/create-relocatable-package.py --mode dev 'build/dev/dist/tar/scylla-unstripped-5.3.0~dev-0.20230505.c356c0c8d376.x86_64.tar.gz'
22:17:47  FAILED: build/dev/dist/tar/scylla-unstripped-5.3.0~dev-0.20230505.c356c0c8d376.x86_64.tar.gz 
22:17:47  scripts/create-relocatable-package.py --mode dev 'build/dev/dist/tar/scylla-unstripped-5.3.0~dev-0.20230505.c356c0c8d376.x86_64.tar.gz'
22:17:47  Traceback (most recent call last):
22:17:47    File "/jenkins/workspace/scylla-master/scylla-ci/scylla/scripts/create-relocatable-package.py", line 170, in <module>
22:17:47      ar.reloc_add('build/node_exporter', arcname='node_exporter')
22:17:47    File "/jenkins/workspace/scylla-master/scylla-ci/scylla/scripts/create-relocatable-package.py", line 24, in reloc_add
22:17:47      return self.add(name, arcname="{}/{}".format(RELOC_PREFIX, arcname),
22:17:47             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
22:17:47    File "/usr/lib64/python3.11/tarfile.py", line 2008, in add
22:17:47      self.add(os.path.join(name, f), os.path.join(arcname, f),
22:17:47    File "/usr/lib64/python3.11/tarfile.py", line 2002, in add
22:17:47      self.addfile(tarinfo, f)
22:17:47    File "/usr/lib64/python3.11/tarfile.py", line 2030, in addfile
22:17:47      copyfileobj(fileobj, self.fileobj, tarinfo.size, bufsize=bufsize)
22:17:47    File "/usr/lib64/python3.11/tarfile.py", line 250, in copyfileobj
22:17:47      raise exception("unexpected end of data")
22:17:47  OSError: unexpected end of data
22:17:47  Exception ignored in: <function _Stream.__del__ at 0x7f60a0d39080>
22:17:47  Traceback (most recent call last):
22:17:47    File "/usr/lib64/python3.11/tarfile.py", line 411, in __del__
22:17:47    File "/usr/lib64/python3.11/tarfile.py", line 461, in close
22:17:47  ValueError: write to closed file
```

Started another build.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1075/

avikivity commented on db/view/view.cc:
```diff
@@ -1754,6 +1754,36 @@ future<> view_update_generator::mutate_MV(
+
+    // In case of an ongoing topology change it might also be necessary to send the update to pending nodes.
+    inet_address_vector_topology_change pending_endpoints =
+        _proxy.local().get_token_metadata_ptr()->pending_endpoints_for(view_token, keyspace_name);
```
The function should get an effective_replication_map_ptr from its caller and use that. The caller should keep the effective_replication_map_ptr captured until it stops using the results of this calculation.

The reason is that the calculation can be outdated as soon as the function returns. Capturing the effective_replication_map_ptr keeps track of how many users reference the older topology, and later we'll use it as a barrier to prevent the topology state machine from moving forward.

/cc @bhalevy

cvybhu replied:
I read up a bit about topology changes (mostly just https://github.com/scylladb/scylladb/blob/master/docs/dev/topology-over-raft.md), AFAIU the gist of it is:

When a new node is added it takes over some token ranges. While it's joining the cluster we send the updates to both the old owner of the token range and the new node. Once the joining is finished we switch over to the new node completely.

There is a risk of the following race:
* We have a view update to send, from the current topology we deduce it should be sent to node A
* While the update is being sent a topology change happens and now it should be sent to node B instead
* The update is sent to the wrong node, view inconsistency happens

To avoid this each update should get an instance of `effective_replication_map_ptr` which prevents the switch to B until the change has fully completed, and it should keep this `ptr` until the update has completed fully.

If this understanding is correct, then the current view update sending is buggy, as it doesn't use the `effective_replication_map_ptr` at all:
https://github.com/scylladb/scylladb/blob/20f620feb9a26ca0d7a78018fc79ac522a52513a/db/view/view.cc#L1639-L1640

I guess it should be changed, although I'm not sure if it belongs in this PR.

cvybhu replied:
Regarding the new code - maybe the throttling calculation doesn't really need to lock the topology? It's just a heuristic, information about remote backlogs isn't 100% up to date and the calculation doesn't need to be 100% exact, it's just an approximation of how much work the other nodes are doing.

There's a risk that comes with locking the topology for the whole duration of a throttled operation. We do a sleep that could last arbitrarily long, possibly even many seconds. In the usual case the topology is only locked for the amount of time required to send a single request, but here the lock might be a lot longer. Can the topology code handle being locked for a whole minute? Or will timeouts start popping up?

I guess it's also an issue with normal view updates, even without the throttling. If we lock the topology at the point of generating a view update, then the lock will be kept for the amount of time that the update is in the queue. The queue of view updates can take up 10% of the available memory, so there could be a significant delay between generating the view update and actually sending it. This could be fixed by locking the topology at the moment of sending, but that would require a bit of refactoring.

Overall I think this PR would be better off without any topology locking - taking the lock before sleeping could cause issues, and the throttling calculation doesn't need to be 100% exact, so IMO the pros of not locking outweights the cons.

avikivity commented on db/view/view.cc:
```diff
@@ -1754,6 +1754,36 @@ future<> view_update_generator::mutate_MV(
+    for (const gms::inet_address& pending_endpoint : pending_endpoints) {
+        // std::find is there to avoid any possible duplicates.
+        if (std::find(result.begin(), result.end(), pending_endpoint) == result.end()) {
+            result.push_back(pending_endpoint);
```
There is contains(). But why do you expect duplicates?

cvybhu replied:
The original `mutate_MV` had some comments about duplicates so I added a bit of code to make sure that there are none, it doesn't really cost anything on a 2-element vector.

avikivity commented on service/storage_proxy.cc:
```diff
@@ -1381,19 +1381,19 @@ class abstract_write_response_handler : public seastar::enable_shared_from_this<
+        // at the timeout point anyway.
+        std::chrono::nanoseconds budget =
+            std::max(std::chrono::nanoseconds(0), _expire_timer.get_timeout() - storage_proxy::clock_type::now());
+
```
Why use nanoseconds instead of duration_type?

I don't understand the problem the patch is fixing.

cvybhu replied:
The original code looked like this:
```c++
std::chrono::microseconds calculate_delay(db::view::update_backlog backlog) {
    constexpr auto delay_limit_us = 1000000;
    auto adjust = [] (float x) { return x * x * x; };
    auto budget = std::max(storage_proxy::clock_type::duration(0),
        _expire_timer.get_timeout() - storage_proxy::clock_type::now());
    std::chrono::microseconds ret(uint32_t(adjust(backlog.relative_size()) * delay_limit_us));
    // "budget" has millisecond resolution and can potentially be long
    // in the future so converting it to microseconds may overflow.
    // So to compare buget and ret we need to convert both to the lower
    // resolution.
    if (std::chrono::duration_cast<storage_proxy::clock_type::duration>(ret) < budget) {
        return ret;
    } else {
        // budget is small (< ret) so can be converted to microseconds
        return std::chrono::duration_cast<std::chrono::microseconds>(budget);
    }
}
```
`storage_proxy::clock_type::duration` is an alias for `std::chrono::nanoseconds`.

The original code assumed that `storage_proxy::clock_type::duration` is `std::chrono::millisconds`, not `nanoseconds`.
It calculates the amount of time to sleep for, the result is `std::chrono::microseconds`.
There's also a `budget` - maximum amount of time that we are allowed to sleep for. Budget is expressed as `storage_proxy::clock_type::duration`, which the original code assumed to be `milliseconds`, but is actually `nanoseconds`.

Directly comparing `milliseconds` and `microseconds` is risky - casting `int64::max()` of milliseconds to microseconds would lead to overflow, as the cast multiplies the amount of milliseconds by 1000.

To avoid overflow the original code did the comparison in a safe way:
```c++
(milliseconds)ret < budget
```

Casting the ret, which is in `microseconds` to budget's type, `milliseconds` is safe - this is a division by 1000.

But the problem is that the budget is actually in `nanoseconds`, not `milliseconds`.
Because of this the anti-overflow comparison becomes:
```c++
(nanoseconds)ret < budget
```
This is unsafe - ret is in `microseconds`, so casting it to `nanoseconds` means multiplying by 1000, which could overflow.

The commit fixes the issue by casting the `budget` from `nanoseconds` to `microseconds` (division by 1000, safe) and then doing the comparison.

I used the `std::chrono::nanoseconds` directly instead of the alias to signify that the value is expected to be in `nanoseconds`. If somebody changes `storage_proxy::clock_type::duration` to be something else this part of code won't compile and they will have to take a look at this part of the code and modify overflow protection to handle the new type.

avikivity replied:
Please extract it into a separate pull request with an issue if you think it's fixing a bug. We also have to think about how this is tested.

avikivity replied:
But I don't see how the original code can fail. It's an expiration time. Even if many years, it will fit in 64 bits.

cvybhu replied:
> Please extract it into a separate pull request with an issue if you think it's fixing a bug. We also have to think about how this is tested.

Opened https://github.com/scylladb/scylladb/pull/14637

avikivity commented on replica/table.cc:
```diff
@@ -2039,6 +2072,29 @@ future<> table::generate_and_propagate_view_updates(shared_ptr<db::view::view_up
+                std::chrono::microseconds throttle_delay =
+                    calculate_throttling_delay_for_view_update_batch(*updates, base_token, gen, timeout);
+
+                co_await seastar::sleep(std::min(throttle_delay, throttle_delay));
```
What's the point of std::min(x, x)?

cvybhu replied:
Oops a leftover, will remove.

avikivity commented:
We'll somehow need to test this. But test.py or dtest aren't any good for performance sensitive tests.

avikivity commented:
Let's say that we delete a partition with a billion rows. Will we then reject or indefinitely delay further updates to the base table, since the backlog is so high?

denesb commented on db/view/view_update_generator.hh:
```diff
@@ -75,6 +76,8 @@ public:
 
     replica::database& get_db() noexcept { return _db; }
 
+    const sharded<service::storage_proxy>& get_storage_proxy() const noexcept {return _proxy;};
```
Why const? The storage proxy is not part of the state of the view update generator, so it is fine to return mutable reference.

cvybhu replied:
I didn't modify anything so I added the const, but I didn't think about it much, will remove.

denesb commented on exceptions/exceptions.hh:
```diff
@@ -144,6 +144,12 @@ struct mutation_write_timeout_exception : public request_timeout_exception {
 
+// Generating view updates for a single client request can take a long time and might not finish before the timeout is
+// reached. In such case this exception is thrown.
+struct view_update_generation_timeout_exception : public cassandra_exception {
```
How is this exception used? We have deliberate mechanisms on the read and write paths to recognize and handle timeout errors and this exception will not be recognized as such. Maybe make `request_timeout_exception` its base class.

cvybhu replied:
Alright I moved things around and now `view_update_generation_timeout_exception` derives from `request_timeout_exception`.

denesb commented on replica/table.cc:
```diff
@@ -1994,6 +1998,32 @@ static size_t memory_usage_of(const utils::chunked_vector<frozen_mutation_and_sc
+    // Find the maximum batchlog, initial value is the local batchlog
+    db::view::update_backlog max_backlog = gen->get_storage_proxy().local().get_view_update_backlog();
+
+    for (const frozen_mutation_and_schema& view_update : view_updates) {
```
How many of these can we have? `chunked_vector` suggests that his can be potentially large, which makes the run-time of this function a concern, especially w.r.t. stalls.

cvybhu replied:
A single batch of updates can contain at most 100 rows worth of updates. This means that the maximum length of this vector is 100, in the case of 100 single-row updates.

100 isn't a lot, the number was chosen to be small enough so that handling all of them at once wouldn't cause stalls.

Potentially we could have a problem if the topology lookup took a long time, but I doubt it, it's in the hot path as it's being called on every update.

denesb replied:
I don't think the topology lookup is a concern, I was only concerned by the number of elements we have to process here, but if the max is 100 (or that order of magnitude) we should be fine.

denesb reviewed (commented):
The approach looks good to me.

We also need @nyh to review.

cvybhu commented:
v2:
* Fix the useless `std::min`
* Make `view_update_generation_timeout_exception` inherit from `request_timeout_exception`.
* Add a comment saying why the throttling calculation doesn't require topology locking.

cvybhu commented:
I took a closer look at the view update backlogs, and I think I see one more risk here.

A view update backlog of node X is the amount of pending view updates generated on node X. When a view update is generated it eats some units from the local node's view update semaphore and that counts in the backlog. Once the update is sent, we receive the response and all is good then the units are freed.

For throttling we take a look at the local node's backlog and the backlogs of nodes to which the update will be sent.

Information about the remote backlogs is updated on each request, and additionally via gossip.

The problem is that a backlog only counts the number of updates that are generated on some node, not the amount of updates that the node is currently processing.

Imagine the following scenario:
* There are 5 nodes: A, B, C, D, E
* Nodes A, B, C and D receive a `DELETE` request, and it turns out that it only deletes rows that are kept on node E, so all of the view updates are sent to node E
* For throttling, node A looks at backlogs of A and E. A's backlogs might fill up as the updates are generated, but E's backlog will be empty, as it doesn't generate any view updates.
* The same situation repeats on B, C and D. Each of them looks at their own backlog and at E's empty backlog.
* Nodes A, B, C, D generate as many updates as their backlog allows, which might overwhelm E, as E's backlog is always empty and it doesn't have a way to slow down the nodes that are sending the updates.
* The situation gets worse the more nodes there are, 10 nodes will generate 10x as much updates as a single node, E has no way to slow them down.

cvybhu commented:
I guess the risk of this happening is rather low, it's a bit synthetic to have a `DELETE` which deletes rows from a single partition and is done concurrently on multiple nodes.

Throttling is a heuristic, and it might be good enough.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1177/

denesb commented:
> * Nodes A, B, C and D receive a `DELETE` request, and it turns out that it only deletes rows that are kept on node E, so all of the view updates are sent to node E

Doesn't each base replica send updates to just one paired view replica. So in this case, only one node would be sending updates for the `DELETE` to `E`, regardless of how many other base replicas there are.

cvybhu commented:
I ran two tests to see that this really works as expected.
I put them here: https://github.com/cvybhu/rusty-scylla-tests

## Test 1
The first test was a port of the python dtest described in this issue.
I ported it to rust to avoid troubles with running dtests locally.

It does a single `DELETE WHERE pk IN (0, ..., 49)`.
Before the change this crashed the cluster, after the change everything works correctly.

## Test 2
The second test is synthetically designed to be as hostile for the throttling algorithm as possible.

It follows the scenario described in https://github.com/scylladb/scylladb/pull/13792#issuecomment-1541976158

There is a table and a bunch of materialized views:
```cql
CREATE KEYSPACE ks WITH replication = {{'class': 'SimpleStrategy', 'replication_factor': {}}}
CREATE TABLE tab (pk bigint, ck bigint, r bigint, PRIMARY KEY (pk, ck))
CREATE MATERIALIZED VIEW {view_name} AS SELECT pk, ck, r FROM tab WHERE pk IS NOT NULL and ck IS NOT NULL and r IS NOT NULL PRIMARY KEY (r, ck, pk)
...
```

There are N nodes in the cluster. The values for `pk` are chosen so that they're stored on nodes `1, ..., (N-1)`.
The value for `r` is chosen so that it belongs to node `N`, all materialized view rows belong to a single partition that's stored on this node.

This creates a scenario in which base table updates are handled by nodes `1, ..., (N-1)`, but all of them generate remote view updates that are sent to node `N`. `N`'s backlog is empty, no matter how many view updates it receives, so it could potentially get overloaded.

The test first inserts a bunch of rows and then deletes all of them at once.

My worry was that this would could an overload, even with the throttling upgrade.
Since every node only cares about its own backlog, the number of pending view updates would increase linearly with the number of nodes in the cluster.
This didn't really happen, even on a 13 node cluster. All of the nodes survived without crashes.
My theory is that the OOM overload can only be caused locally, for remote view updates there's some kind of limitating factor (something in storage_proxy maybe?). Looking at the local backlog saves us from the local overload.

The only overload that I managed to get was during inserts. On the 13 node cluster with 32 views trying to perform inserts
at a concurrency higher than 32 caused errors (received 0 responses, 1 required), even with 1 hour timeouts.
I guess this happens because all of the udpates are sent to the single node.

The scenario is highly artificial, maybe this isn't expected to really happen in nature.

cvybhu commented:
> Let's say that we delete a partition with a billion rows. Will we then reject or indefinitely delay further updates to the base table, since the backlog is so high?

I think a deletion can be interpreted as a writer who does 100 concurrent inserts. It does 100 requests, then gets throttled before doing another 100.
This is somewhat reasonable. A delete is heavy operation and can slow others down, but a single `DELETE` and 100 parallel inserts would theoretically split the bandwidth equally between themselves.
Maybe it would be good to have an actual smart scheduler there, but this sounds good enough.

cvybhu commented:
v2.1
* Rebased onto the latest master, there were conflicts because of 87307781c43914b0dbb7503908100371de227800 (different function for `get_pending_endpoints`) and 66e43912d6f0d82ce465b462d44a42ff76c41925 (removed `io_priority` from `make_reader_v2`)

I think the PR is ready, it needs review and approval.

denesb commented on db/view/view_update_generator.hh:
```diff
@@ -12,6 +12,7 @@
 #include "db/timeout_clock.hh"
 #include "utils/chunked_vector.hh"
 #include "gms/inet_address.hh"
+#include "db/view/view_update_backlog.hh"
```
Why this new include?

denesb commented on db/view/view_update_generator.hh:
```diff
@@ -94,6 +97,7 @@ public:
     utils::small_vector<gms::inet_address, 2> get_endpoints_for_view_update(
         dht::token base_token,
         const frozen_mutation_and_schema& view_update) const;
+
```
Spurious change.

denesb commented on replica/table.cc:
```diff
@@ -2035,6 +2070,29 @@ future<> table::generate_and_propagate_view_updates(shared_ptr<db::view::view_up
+        try {
+            if (batch_num > 0) {
+                std::chrono::microseconds throttle_delay =
+                    calculate_throttling_delay_for_view_update_batch(*updates, base_token, gen, timeout);
```
We already calculate the endpoints the views have to be sent to in `calculate_throttling_delay_for_view_update_batch()`. I don't know how expensive that is, but maybe it is worth trying to avoid duplicating this effort, by calculating once and then pass it to both `calculate_throttling_delay_for_view_update_batch()` and `mutate_MV()`.

nyh replied:
The problem is that the new calculate_throttling_delay_for_view_update_batch() calculates the endpoints for 100 different updates. I think in the general (and common) case it will just look at all the nodes in the cluster, in a rather inefficient way ;-)

cvybhu replied:
Eh I guess it'll most often look at the whole cluster, but is it that inefficient?

There are cases where the view updates are sent to a few nodes at most (e.g local secondary index clustering range delete), In such cases I don't want to involve the backlogs of other nodes in the throttling calculation, as it'd be wasteful.

cvybhu commented:
v3:
* remove needless include
* remove spurious newline change

cvybhu commented:
> I was expecting that this function replaces the code in mutate_MV(), but I see that it doesn't. Why?

I initially tried to do make `mutate_MV` use the new function, but decided not to for two reasons.

1) `mutate_MV` has a very convoluted flow. The division into target endpoints and pending endpoints is deeply entrenched in the logic there. If I wanted it to use the new function I would have to pretty much rewrite it from scratch. I didn't want to add so much needless complexity to this PR, and

2) I think `mutate_MV` should use the `effective_replication_map` instead of `get_view_natural_endpoint`. The current code seems susceptible to race conditions during schema changes. Imagine the following scenario:
  * `mutate_MV` is called, it sees that the update should be sent to `node1`, and schedules the rpc call to be sent there
  * A schema change happens. During this change token ranges change and now the update should be sent to `node2`, not `node1`. `node1` doesn't care about this token range at all
  * The rpc call is sent to `node1` anyway, view inconsistency happens
  
    AFAIU `effective_replication_map` is a tool that can be used to avoid scenarios like this. Holding an instance of it locks schema change for the duration of the update. So I think `mutate_MV` should get an instance of `effective_replication_map`, send the update, receive response, then drop the instance.

So I think that `mutate_MV` should generally be rewritten, but I don't want to block this PR any further. We could consider reusing data from throttling calculation during the rewrite.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1870/

cvybhu commented:
ping, could we merge this?

denesb commented:
> ping, could we merge this?

I was hoping to get another round of review from @nyh too.

avikivity commented:
I'll review too.

avikivity commented on db/view/view.cc:
```diff
@@ -1755,6 +1755,39 @@ future<> view_update_generator::mutate_MV(
+    inet_address_vector_topology_change pending_endpoints = _proxy.local()
+                                                                .local_db()
+                                                                .find_keyspace(keyspace_name)
+                                                                .get_effective_replication_map()
```
It's wrong to use get_effective_replication_map() for a point query. Its results are only valid for as long as the pointer is held.

Correct use it:
 1. at the beginning of the operation, capture the effective replication map
 2. use it for all such queries
 3. at the end of the operation (including acknowledgement from remotes), release the erm_ptr

avikivity replied:
This allows the rest of the system to avoid changing the topology to an incompatible version while the operation is in progress.

cvybhu replied:
Yes, this is a bit "incorrect", but it is a conscious decision.
This function is used only to calculate the throttling delay, which doesn't need to be perfect, it's a heuristic.
There's no need to block the whole schema change for this.

I agree that `mutate_MV` should use the `effective_replication_map` (it doesn't at the moment), but there's more risks there, as each view update has a 5 minute timeout, so a schema change could be blocked for the whole 5 minutes.

avikivity commented on db/view/view_update_generator.hh:
```diff
@@ -75,6 +75,8 @@ public:
 
     replica::database& get_db() noexcept { return _db; }
 
+    const sharded<service::storage_proxy>& get_storage_proxy() const noexcept {return _proxy;};
```
A few spaces would be nice.

avikivity commented on service/storage_proxy.cc:
```diff
@@ -1386,28 +1386,11 @@ class abstract_write_response_handler : public seastar::enable_shared_from_this<
-        std::chrono::microseconds max_delay = std::chrono::duration_cast<std::chrono::microseconds>(budget);
-
-        return std::min(ret, max_delay);
-    }
```
I don't know if we want to reuse the sleep-based throttling here.

For the client, we can't control its concurrency, so we just slow it down.

But for view update generation, isn't it enough to limit it with a semaphore?

nyh replied:
@avikivity, how would you use a semaphore?

Imagine that there is just **one** delete going on, that wants to delete a million rows. After deleting a batch 100 rows, how quickly should the deletion send another batch of 100 rows?  If it sends the next batch immediately, it overloads the server with background work (that's the current situation).

I guess it could somehow wait until these 100 specific view rows were written, but why 100? The idea of the sleep is to use the same flow control algorithm that an ordinary "deletion client" with concurrency 100 would have used. This deletion client may send much more than 100 concurrent writes as a burst, but as soon as the backlog begins to grow, it will start slowing down.

cvybhu replied:
We could limit the amount of concurrent `view updates from range delete` operations, but this causes the usual issues with limiting concurrency - what number of concurrent requests is the right one - 100, 10000, 1e6?. With high latency we want the window size to be large, with low latency it should be small.

This solution treats each range delete as a INSERT workload with concurrency of 100, and uses the current backpressure mechanism to limit the number of concurrent updates.

IMO ideally it should look like this:
* The node that generates the updates pushes them onto a TCP connection, blocking until the `send()` completes.
* The receiver node reads as many updates as it can reasonably handle at once (say, 1000 concurrent requests)
* TCP flow control takes care of choosing the right amount of in-flight requests

But I think this would require some changes to RPC. I have some thoughts about changing how the backpressure works, I'll try to make a clear document about it.

For now I feel that this is the best solution - it uses the existing mechanism and fixes the issue.

avikivity replied:
We need to maintain some concurrency limit. In the worse case, every single shard is performing view updates, so the total concurrency is one_shard_concurrency * nr_shards. In the worst case all of this concurrency is aimed at one shard (which would cause immense slowdowns, but that's life). In the best case, it's spread evenly across the cluster.

I think shard-wide (or service-level-wide) concurrency of 100 is reasonable. And above that we delay the client (by delaying the coordinator).

With sleeping, the concurrency just keeps growing. Yes, we're sending fewer requests, but our own concurrency is unbounded.

avikivity replied:
> > With sleeping, the concurrency just keeps growing. Yes, we're sending fewer requests, but our own concurrency is unbounded.
> 
> Why does our concurrency just keep growing? We're (the base replica processing the partition delete) are sending fewer requests, and not doing anything else. We're sending batches of row deletes 100 at a time, and don't have any background work besides the work we send to other nodes, so how can **our** concurrency be growing? The concurrency on those other nodes can indeed grow with all that background work, but that concurrency we check - and if it grows - we increase the sleep.

We're processing new incoming replica writes at the same time.

Let's say we get a delete from a 1 GB partition. We start the send/sleep/send thing for that partition.
A millisecond later, another 1GB delete.
After a second, we're asked to delete 1000 such partitions. We have a concurrency of 1000.

avikivity replied:
The whole idea of batching is wrong. We should have a concurrency limit, but why batch? read a row, process it (background), read another row. process it (background), eventually you hit the concurrency limit and wait until some row completes.

Would have solved the issues with range tombstones that span batch boundaries too.

avikivity replied:
100ms is not typical latency for a base replica and a view replica.

avikivity replied:
And we can increase 100 to 1000, just as long as it's bounded. Changing from one unbounded algorithm to another isn't good.

cvybhu replied:
The view updates are supposed to be asynchronous - we spawn as many of them as will fit in the queue, and then process them in the background. The only waiting we should be doing is waiting for a place in the queue.
Bounding this to 100 requests goes against this idea, it will make all such view updates synchronous.

avikivity replied:
> > Let's say we get a delete from a 1 GB partition. We start the send/sleep/send thing for that partition. A millisecond later, another 1GB delete. After a second, we're asked to delete 1000 such partitions. We have a concurrency of 1000.
> 
> If the user sends 1000 deletions of 1GB partitions concurrently, there will indeed be a lot of concurrent work - the internal concurrency will be 1000*100. But is this unexpected?

Why not?

> The user knows these are huge deletions, they should expect that doing 1000 of them in parallel should be difficult.

Why would they know that? And why would they care? The database shouldn't crash because you throw something difficult at it.

> Not impossible - concurrency of 100,000 is not out of the question. Anyway, what is the alternative?

Throttle down the incoming base replica work.

> If you have 1000 concurrent partition deletions, and each needs to process 1 million row deletions, would you expect each partition deletion to process and send one row at a time? Today we have the batch size 100 just to make the reading part more efficient, would we want to lower the batch size if there are a lot of concurrent partition deletions? I'm not sure.

No. Limit each service level's view update concurrency to something reasonable (100-1000). Once it's exceeded, require progress to be made on the view updates before processing more base updates.

The batching is wrong, we already have readers that do batching. The extra batching on top is a source of bugs and inefficiency.

avikivity replied:
> The view updates are supposed to be asynchronous - we spawn as many of them as will fit in the queue, and then process them in the background. The "concurrency" is supposed to be sky high. The only waiting we should be doing is waiting for a place in the queue. Bounding this to 100 requests goes against this idea, it will make all such view updates synchronous.


What limits the queue size?

To have end-to-end throttling, we need to stop processing base updates (which will stop processing client requests as the back-pressure propagates).

I don't understand how the sleeps help. They spread the requests over time, but more keep coming in. If our base replica processing throughput is higher than the view update throughput, we'll run out of memory.

nyh replied:
@avikivity you're absolutely right that "the extra batching on top is a source of bugs" - see for example 92d03be37b407dfb48d3cf96ef28bfbe42aed3b2. But this is what happens when we keep piling on fixes on fixes on fixes on fixes without having the time and peace of mind to actually give some TLC to our code. So we already have this 100-row batching in place.

If you're curious where this 100-row-batching came from, it started by reading the whole partition sequentially, and then in bf0777e97a0abb3bebddf84c82de774ebfd9255d this 100-row batches happened.

avikivity commented on replica/table.cc:
```diff
@@ -2035,6 +2070,29 @@ future<> table::generate_and_propagate_view_updates(shared_ptr<db::view::view_up
         }
         tracing::trace(tr_state, "Generated {} view update mutations", updates->size());
         auto units = seastar::consume_units(*_config.view_update_concurrency_semaphore, memory_usage_of(*updates));
+
```
Instead of this consume_units, why not also wait for units to become available?

btw, this should be separate for system/user (and service level) to prevent priority inversion.

cvybhu replied:
There's the idea that view updates shouldn't block normal queries, to maintain high availability, so we shouldn't wait on any semaphores. When there's no resources in the semaphore we should just shed the load. Throttling should make sure that this never happens (in a workload with limited concurrency). I'm not sure how valid this idea is, I tried to change it in another PR, but there was some backpressure against this idea.

/cc @denesb

avikivity replied:
Once the server is overloaded, we have two options: throttle normal queries, or drop view updates. Of the two, throttling normal queries is much better.

Just sleeping without throttling the real client will mean we'll have more and more sleepers, until this node's memory is consumed by sleeping view updates.

cvybhu replied:
The real client is throttled as well because the view backlog increases as the view updates are generated.
With large backlog, all normal client writes will be throttled as well.

The idea of sleep-based throttling is that it will eventually sleep enough to ensure that the amount of view updates doesn't increase anymore. The concurrency is bounded, but in an indirect way.

avikivity replied:
Can we prove that this indeed happens?

nyh commented on exceptions/exceptions.cc:
```diff
@@ -70,6 +70,10 @@ read_write_timeout_exception::read_write_timeout_exception(exception_code code,
 
+view_update_generation_timeout_exception::view_update_generation_timeout_exception()
+    : request_timeout_exception{
+          exception_code::WRITE_TIMEOUT, "Request timed out - couldn't generate materialized view updates in time"} {}
```
Here and in the comment for this class below, it is confusing what "generating" view updates mean. Does it include waiting for the remote view replica to answer? It should be explained.

cvybhu replied:
I changed the comment, and the error message is now:
```
Request timed out - couldn't prepare materialized view updates in time
```
`prepare` signifies that it isn't about sending, the timeout happened earlier, at the "preparation" stage

nyh replied:
But why would the preparation phase take a long time? Were we waiting for the read-before-write?

nyh reviewed (approved):
I marked "approve", but I think that @avikivity and @denesb left some points they want addressed. I also left a few comments, but I don't think anything is critical.

cvybhu commented on db/view/view_update_generator.hh:
```diff
@@ -85,6 +86,14 @@ public:
+    // ongoing topology change it might also be necessary to send the update to pending nodes.
+    utils::small_vector<gms::inet_address, 2> get_endpoints_for_view_update(
+        dht::token base_token,
+        const frozen_mutation_and_schema& view_update) const;
```
answered in https://github.com/scylladb/scylladb/pull/13792#issuecomment-1593790888

cvybhu commented:
v4:
* Added spaces in `get_storage_proxy`
* Adjusted comment on `calculate_view_update_throttling_delay` to be shorter, and added a link to the article
* Adjusted comment and error message of `view_update_generation_timeout_exception` to make it clearer what it means
* Moved the controversial fix of calculate_delay() to a separate PR: https://github.com/scylladb/scylladb/pull/14637. It conflicts a bit with this PR, but I can rebase one of them later without a problem.

cvybhu commented:
I think what we're trying achieve is something like this:

![image](https://github.com/scylladb/scylladb/assets/36861778/c71ff406-8960-4a3f-a581-47695c4c7846)

* There is a queue (imaginary, but queue) of view updates that should be sent to the view replica
* Each base table update that wants to generate view updates should call (imaginary) `queue.push_view_update()`
* `queue.push_view_update()` uses throttling - the more full the queue is the longer `push()` takes to complete. This ensures gentle backpressure.
* View updates are asynchronous - `push()` only enqueues the update to be sent, it doesn't wait for the response. The amount of view updates waiting in the queue can be very large, but this is expected, it isn't a concurrency that should be bounded.
* All view updates are sent over RPC to the receiving node
* The receiving node should receive some requests, and process them. The amount of requests that it processes at once should be bounded, e.g max 1000 view updates at once.

cvybhu commented:
There should probably be a separate queue for each service level, asynchronous updates, synchronous updates, maybe also for the updates that are generated by range deletes.

cvybhu commented:
This PR tries to extend DELETE queries to also throttle on each `push()`, so that the queue doesn't overflow.

We could maybe add an explicit function `enqueue_view_update()` so that all of this logic is in one place

cvybhu commented:
The receiver side might need some modifications. I looked at the RPC server and there's a semaphore that limits the number of concurrent requests, but it looks like it calculates resource usage by looking at the size of the RPC frame. This can cause a lot of view updates to be handled at once, which could maybe cause OOM.

I think it would be good to add a separate semaphore that limits the number of concurrent requests to a limited number of requests, or just uses a better estimate for the amount of memory used.

cvybhu commented:
Although we would have to be careful about limiting the RPC server, because we still use the `MUTATION` verb for view updates.
Imagine this:
* Nodes receive a lot of writes, they don't accept any more
* To complete the writes we need to send view updates, but they can't be sent because nodes don't accept `MUTATION` verbs anymore
* Deadlock happens

cvybhu commented:
Ehh there's a lot of work to do. I'll try to make a doc with some thoughts.

This PR isn't perfect, but it fixes the immediate problem without trying to rewrite everything.

mykaul commented:
> This PR isn't perfect, but it fixes the immediate problem without trying to rewrite everything.

Good path forward - as long as it doesn't introduce regressions, I'm in favor.

scylladb-promoter commented:
CI state **`ABORTED`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2398/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2433/

avikivity commented:
> Although we would have to be careful about limiting the RPC server, because we still use the `MUTATION` verb for view updates. Imagine this:
> 
> * Nodes receive a lot of writes, they don't accept any more
> * To complete the writes we need to send view updates, but they can't be sent because nodes don't accept `MUTATION` verbs anymore
> * Deadlock happens

Well, we should use a separate verb for the view update, so it gets accounted separately.

I think this needs more extensive testing. And we should schedule a more complete fix (that starts with some design work rather than working within the limitations of what we have now).

mykaul commented:
@kostja , @eliransin - do you know what's the status of this one? It certainly missed 5.4. Is 6.0 realistic here?

kostja commented:
It's not going to be worked on until sometime after February.

wmitros commented:
Replaced by https://github.com/scylladb/scylladb/pull/16819
